File: frontend/utils/export.ts
import { Metadata, BatchItem } from '../types';

const downloadFile = (blob: Blob, fileName: string) => {
    const link = document.createElement('a');
    link.href = URL.createObjectURL(blob);
    link.download = fileName;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(link.href);
};

// ── CSV COLUMN DEFINITIONS ────────────────────────────────────────────────────
// Single source of truth for export schema — stays in sync with backend ALLOWED_METADATA_KEYS

const CSV_HEADERS = [
    // Identity
    '_filename',
    'title',
    'artist',
    'album',
    'albumartist',
    'year',
    'track',
    // Sonic / Technical
    'MAIN_INSTRUMENT',
    'TKEY',
    'MODE',
    'bpm',
    'LANGUAGE',
    // Genre & Classification
    'genre',
    'SUBGENRE',
    'MOOD',
    'INSTRUMENTATION',
    'USAGE',
    'KEYWORDS',
    // Description
    'comment',
    // Vocal Style
    'VOCAL_STYLE_GENDER',
    'VOCAL_STYLE_TIMBRE',
    'VOCAL_STYLE_DELIVERY',
    'VOCAL_STYLE_EMOTIONALTONE',
    // Credits & Legal
    'copyright',
    'publisher',
    'composer',
    'lyricist',
    'producer',
    'CATALOGNUMBER',
    'isrc',
    'iswc',
    'upc',
    'P_LINE',
    'SHA256',
    'LICENSE',
    'EXPLICIT_CONTENT',
    'FILE_OWNER',
    // Extended AI fields
    'ENERGY_LEVEL',
    'MOOD_VIBE',
    'MUSICAL_ERA',
    'PRODUCTION_QUALITY',
    'DYNAMICS',
    'TARGET_AUDIENCE',
    'TEMPO_CHARACTER',
    'ANALYSIS_REASONING',
    // DSP metrics
    'DYNAMIC_RANGE_DB',
    'SPECTRAL_CENTROID_HZ',
    'SPECTRAL_ROLLOFF_HZ',
    'ACOUSTIC_SCORE',
    'HAS_VOCALS',
];

const sanitizeScalar = (value: unknown): string => {
    if (value === undefined || value === null) return '';
    return String(value).replace(/\r?\n/g, ' ').trim();
};

const serializeArray = (arr: unknown): string => {
    if (!Array.isArray(arr)) return '';
    const cleaned = arr
        .map(v => sanitizeScalar(v))
        .filter(Boolean);
    const unique = Array.from(new Set(cleaned)).sort((a, b) => a.localeCompare(b));
    return unique.join(', ');
};

const formatCsvValue = (value: unknown): string => {
    const v = sanitizeScalar(value);
    return `"${v.replace(/"/g, '""')}"`;
};

const resolveVocalStyle = (data: Metadata) => {
    const vs = data.vocalStyle;
    const gender = sanitizeScalar(vs?.gender).toLowerCase();
    const isNone = !gender || gender === 'none' || gender === 'instrumental';
    if (isNone) {
        return { gender: 'none', timbre: 'none', delivery: 'none', emotionalTone: 'none' };
    }
    return {
        gender:       sanitizeScalar(vs?.gender),
        timbre:       sanitizeScalar(vs?.timbre),
        delivery:     sanitizeScalar(vs?.delivery),
        emotionalTone: sanitizeScalar(vs?.emotionalTone),
    };
};

const metadataToCsvRow = (fileName: string, data: Metadata): string => {
    if (!fileName) throw new Error('EXPORT_ABORT: missing _filename');

    const vocal = resolveVocalStyle(data);
    const energyLevel = data.energyLevel || data.energy_level || '';

    const values = [
        // Identity
        fileName,
        data.title,
        data.artist,
        data.album,
        data.albumArtist,
        data.year,
        data.track ?? 1,
        // Sonic / Technical
        data.mainInstrument,
        data.key,
        data.mode,
        data.bpm,
        data.language,
        // Genre & Classification
        data.mainGenre,
        serializeArray(data.additionalGenres),
        serializeArray(data.moods),
        serializeArray(data.instrumentation),
        serializeArray(data.useCases),
        serializeArray(data.keywords),
        // Description
        data.trackDescription,
        // Vocal Style
        vocal.gender,
        vocal.timbre,
        vocal.delivery,
        vocal.emotionalTone,
        // Credits & Legal
        data.copyright,
        data.publisher,
        data.composer,
        data.lyricist,
        data.producer,
        data.catalogNumber,
        data.isrc,
        data.iswc,
        data.upc,
        data.pLine,
        data.sha256,
        data.license,
        data.explicitContent,
        data.fileOwner,
        // Extended AI fields
        energyLevel,
        data.mood_vibe,
        data.musicalEra,
        data.productionQuality,
        data.dynamics,
        data.targetAudience,
        data.tempoCharacter,
        data.analysisReasoning,
        // DSP metrics
        data.dynamicRange,
        data.spectralCentroid,
        data.spectralRolloff,
        data.acousticScore,
        data.hasVocals,
    ];

    return values.map(formatCsvValue).join(',');
};


// ── EXPORTS ───────────────────────────────────────────────────────────────────

export const exportToCsv = (data: Metadata, fileName: string) => {
    const row = metadataToCsvRow(fileName || 'track', data);
    const bom = '\uFEFF'; // UTF-8 BOM for Excel compatibility
    const csvContent = bom + [CSV_HEADERS.join(','), row].join('\n');
    const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' });
    downloadFile(blob, fileName.replace(/\.[^/.]+$/, '') + '_metadata.csv');
};


export const exportBatchToCsv = async (batch: BatchItem[]) => {
    const completedItems = batch.filter(item => item.status === 'completed' && item.metadata);
    if (completedItems.length === 0) return;

    // Try backend batch export first (most accurate for server-processed files)
    const jobIds = completedItems.map(item => item.jobId).filter(Boolean) as string[];
    if (jobIds.length > 0 && jobIds.length === completedItems.length) {
        try {
            const { getFullUrl } = await import('../apiConfig');
            const response = await fetch(getFullUrl('/export/batch/csv'), {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(jobIds),
            });
            if (response.ok) {
                const blob = await response.blob();
                downloadFile(blob, `batch_metadata_${Date.now()}.csv`);
                return;
            }
        } catch (e) {
            console.warn('[export] Backend batch export failed, using local fallback:', e);
        }
    }

    // Local fallback
    const csvRows = completedItems.map(item => metadataToCsvRow(item.file.name, item.metadata!));
    const bom = '\uFEFF';
    const csvContent = bom + [CSV_HEADERS.join(','), ...csvRows].join('\n');
    const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' });
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    downloadFile(blob, `metadata_batch_${timestamp}.csv`);
};

File: backend/app/routes/export.py
"""
Export Routes - CSV, JSON, and other format exports
"""

from fastapi import APIRouter, HTTPException, Depends
from fastapi.responses import StreamingResponse, Response
from sqlalchemy.orm import Session
from typing import Dict, Any, List
from datetime import datetime
import csv
import json
import io
import logging

from app.db import SessionLocal, Job

router = APIRouter(prefix="/export", tags=["export"])
logger = logging.getLogger(__name__)


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def metadata_to_csv_row(metadata: Dict[str, Any], filename: str, ipfs_hash: str = None, ipfs_url: str = None) -> Dict[str, str]:
    """
    Convert metadata dictionary to CSV row compatible with MP3Tag.
    
    Format matches mp3tag_mapping.txt specification.
    """
    # Extract vocal style components
    vocal_style = metadata.get('vocalStyle', {})
    
    return {
        'filename': filename,
        'title': metadata.get('title', ''),
        'artist': metadata.get('artist', ''),
        'album': metadata.get('album', ''),
        'year': str(metadata.get('year', '')),
        'track': str(metadata.get('track', '')),
        'MAIN_INSTRUMENT': metadata.get('mainInstrument', ''),
        'TKEY': metadata.get('key', ''),
        'KEY_EXT': metadata.get('key', ''),
        'MODE': metadata.get('mode', ''),
        'bpm': str(metadata.get('bpm', '')),
        'BPM_EXT': str(metadata.get('bpm', '')),
        'genre': metadata.get('mainGenre', ''),
        'SUBGENRE': ', '.join(metadata.get('additionalGenres', [])),
        'comment': metadata.get('trackDescription', ''),
        'KEYWORDS': ', '.join(metadata.get('keywords', [])),
        'copyright': metadata.get('copyright', ''),
        'publisher': metadata.get('publisher', ''),
        'composer': metadata.get('composer', ''),
        'lyricist': metadata.get('lyricist', ''),
        'albumartist': metadata.get('albumArtist', ''),
        'CATALOGNUMBER': metadata.get('catalogNumber', ''),
        'isrc': metadata.get('isrc', ''),
        'MOOD': ', '.join(metadata.get('moods', [])),
        'MOOD_VIBE': metadata.get('mood_vibe', ''),
        'ENERGY': metadata.get('energy_level', ''),
        'INSTRUMENTATION': ', '.join(metadata.get('instrumentation', [])),
        'VOCAL_STYLE_GENDER': vocal_style.get('gender', ''),
        'VOCAL_STYLE_TIMBRE': vocal_style.get('timbre', ''),
        'VOCAL_STYLE_DELIVERY': vocal_style.get('delivery', ''),
        'VOCAL_STYLE_EMOTIONALTONE': vocal_style.get('emotionalTone', ''),
        'USAGE': ', '.join(metadata.get('useCases', [])),
        'LANGUAGE': metadata.get('language', ''),
        'SHA256': metadata.get('sha256', ''),
        'ipfs_hash': ipfs_hash or '',
        'ipfs_url': ipfs_url or '',
    }


@router.get("/csv/{job_id}")
async def export_csv(job_id: str, db: Session = Depends(get_db)):
    """
    Export analysis results to CSV format (MP3Tag compatible).
    
    Returns CSV file for download.
    """
    try:
        job = db.query(Job).filter(Job.id == job_id).first()
        if not job:
            raise HTTPException(status_code=404, detail="Job not found")
        
        if job.status != "completed":
            raise HTTPException(status_code=400, detail="Job not completed yet")
        
        if not job.result:
            raise HTTPException(status_code=400, detail="No results available")
        
        # Convert metadata to CSV row
        csv_row = metadata_to_csv_row(
            job.result, 
            job.file_name,
            getattr(job, "ipfs_hash", None),
            getattr(job, "ipfs_url", None)
        )
        
        # Create CSV in memory
        output = io.StringIO()
        fieldnames = [
            'filename', 'title', 'artist', 'album', 'year', 'track',
            'MAIN_INSTRUMENT', 'TKEY', 'KEY_EXT', 'MODE', 'bpm', 'BPM_EXT', 
            'genre', 'SUBGENRE', 'comment', 'KEYWORDS', 'copyright', 
            'publisher', 'composer', 'lyricist', 'albumartist', 'CATALOGNUMBER', 
            'isrc', 'MOOD', 'MOOD_VIBE', 'ENERGY', 'INSTRUMENTATION', 
            'VOCAL_STYLE_GENDER', 'VOCAL_STYLE_TIMBRE', 'VOCAL_STYLE_DELIVERY', 
            'VOCAL_STYLE_EMOTIONALTONE', 'USAGE', 'LANGUAGE', 'SHA256',
            'ipfs_hash', 'ipfs_url'
        ]
        
        writer = csv.DictWriter(output, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerow(csv_row)
        
        # Prepare response
        output.seek(0)
        filename = f"metadata_{job.file_name.rsplit('.', 1)[0]}.csv"
        
        return StreamingResponse(
            io.BytesIO(output.getvalue().encode('utf-8')),
            media_type="text/csv",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"CSV export failed for job {job_id}: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/json/{job_id}")
async def export_json(job_id: str, pretty: bool = True, db: Session = Depends(get_db)):
    """
    Export analysis results to JSON format.
    
    Args:
        job_id: Job ID
        pretty: Pretty-print JSON (default: True)
    
    Returns JSON file for download.
    """
    try:
        job = db.query(Job).filter(Job.id == job_id).first()
        if not job:
            raise HTTPException(status_code=404, detail="Job not found")
        
        if job.status != "completed":
            raise HTTPException(status_code=400, detail="Job not completed yet")
        
        if not job.result:
            raise HTTPException(status_code=400, detail="No results available")
        
        # Create JSON structure
        export_data = {
            "metadata": job.result,
            "file_info": {
                "original_filename": job.file_name,
                "analysis_timestamp": job.timestamp.isoformat() if job.timestamp else None,
                "job_id": job.id
            }
        }
        
        # Serialize to JSON
        if pretty:
            json_str = json.dumps(export_data, indent=2, ensure_ascii=False)
        else:
            json_str = json.dumps(export_data, ensure_ascii=False)
        
        # Prepare response
        filename = f"metadata_{job.file_name.rsplit('.', 1)[0]}.json"
        
        return Response(
            content=json_str,
            media_type="application/json",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"JSON export failed for job {job_id}: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/batch/csv")
async def export_batch_csv(job_ids: List[str], db: Session = Depends(get_db)):
    """
    Export multiple analysis results to a single CSV file.
    
    Args:
        job_ids: List of job IDs to export
    
    Returns CSV file with all results.
    """
    try:
        # Fetch all jobs
        jobs = db.query(Job).filter(Job.id.in_(job_ids)).all()
        
        if not jobs:
            raise HTTPException(status_code=404, detail="No jobs found")
        
        # Create CSV in memory
        output = io.StringIO()
        fieldnames = [
            'filename', 'title', 'artist', 'album', 'year', 'track',
            'MAIN_INSTRUMENT', 'TKEY', 'KEY_EXT', 'MODE', 'bpm', 'BPM_EXT', 
            'genre', 'SUBGENRE', 'comment', 'KEYWORDS', 'copyright', 
            'publisher', 'composer', 'lyricist', 'albumartist', 'CATALOGNUMBER', 
            'isrc', 'MOOD', 'MOOD_VIBE', 'ENERGY', 'INSTRUMENTATION', 
            'VOCAL_STYLE_GENDER', 'VOCAL_STYLE_TIMBRE', 'VOCAL_STYLE_DELIVERY', 
            'VOCAL_STYLE_EMOTIONALTONE', 'USAGE', 'LANGUAGE', 'SHA256'
        ]
        
        writer = csv.DictWriter(output, fieldnames=fieldnames)
        writer.writeheader()
        
        # Write each job's metadata
        for job in jobs:
            if job.status == "completed" and job.result:
                csv_row = metadata_to_csv_row(job.result, job.file_name)
                writer.writerow(csv_row)
        
        # Prepare response
        output.seek(0)
        filename = f"batch_metadata_{len(jobs)}_tracks.csv"
        
        return StreamingResponse(
            io.BytesIO(output.getvalue().encode('utf-8')),
            media_type="text/csv",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Batch CSV export failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/batch/json")
async def export_batch_json(job_ids: List[str], pretty: bool = True, db: Session = Depends(get_db)):
    """
    Export multiple analysis results to a single JSON file.
    
    Args:
        job_ids: List of job IDs to export
        pretty: Pretty-print JSON (default: True)
    
    Returns JSON file with all results.
    """
    try:
        # Fetch all jobs
        jobs = db.query(Job).filter(Job.id.in_(job_ids)).all()
        
        if not jobs:
            raise HTTPException(status_code=404, detail="No jobs found")
        
        # Create JSON structure
        export_data = {
            "tracks": [],
            "summary": {
                "total_tracks": len(jobs),
                "completed": sum(1 for j in jobs if j.status == "completed"),
                "export_timestamp": datetime.utcnow().isoformat() + "Z"
            }
        }
        
        # Add each job's metadata
        for job in jobs:
            if job.status == "completed" and job.result:
                export_data["tracks"].append({
                    "filename": job.file_name,
                    "metadata": job.result,
                    "analysis_timestamp": job.timestamp.isoformat() if job.timestamp else None
                })
        
        # Serialize to JSON
        if pretty:
            json_str = json.dumps(export_data, indent=2, ensure_ascii=False)
        else:
            json_str = json.dumps(export_data, ensure_ascii=False)
        
        # Prepare response
        filename = f"batch_metadata_{len(jobs)}_tracks.json"
        
        return Response(
            content=json_str,
            media_type="application/json",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Batch JSON export failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))
@router.get("/ddex/{job_id}")
async def export_ddex(job_id: str, db: Session = Depends(get_db)):
    """
    Export analysis results to DDEX ERN 4.3 XML format.
    """
    try:
        job = db.query(Job).filter(Job.id == job_id).first()
        if not job:
            raise HTTPException(status_code=404, detail="Job not found")
        
        if job.status != "completed":
            raise HTTPException(status_code=400, detail="Job not completed yet")
        
        if not job.result:
            raise HTTPException(status_code=400, detail="No results available")
        
        from app.services.ddex_orchestrator import DDEXOrchestrator
        from app.services.sonic_intelligence import TrackMetadata
        
        # Prepare metadata for orchestrator
        # We need to map the stored result to TrackMetadata schema
        metadata = job.result
        vocal_style = metadata.get('vocalStyle', {})
        
        track_data = TrackMetadata(
            title=metadata.get('title') or "Unknown Title",
            artist=metadata.get('artist') or "Unknown Artist",
            isrc=metadata.get('isrc') or "US-XXX-YY-00001",
            bpm=float(metadata.get('bpm') or 0),
            key=f"{metadata.get('key', '')} {metadata.get('mode', '')}".strip(),
            lufs=-14.0, # Result might not have LUFS yet
            danceability=0.5,
            mood_vibe=metadata.get('mood_vibe') or "Neutral",
            energy_level=0.5,
            fingerprint=metadata.get('sha256') or "0"*64
        )
        # Note: TrackMetadata Pydantic model might validate fields strictness. 
        # Duration handling logic is inside Orchestrator template currently as default. 
        
        xml_content = DDEXOrchestrator.generate_xml(track_data)
        
        filename = f"ddex_{job.file_name.rsplit('.', 1)[0]}_{job_id[:8]}.xml"
        
        return Response(
            content=xml_content,
            media_type="application/xml",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"DDEX export failed for job {job_id}: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/cwr/{job_id}")
async def export_cwr(job_id: str, db: Session = Depends(get_db)):
    """
    Export analysis results to CWR (Common Works Registration) V2.1 format.
    """
    try:
        job = db.query(Job).filter(Job.id == job_id).first()
        if not job:
            raise HTTPException(status_code=404, detail="Job not found")
        
        if job.status != "completed":
            raise HTTPException(status_code=400, detail="Job not completed yet")
        
        if not job.result:
            raise HTTPException(status_code=400, detail="No results available")
        
        from app.services.cwr_gen import CWRGenerator
        
        # Generate CWR content
        cwr_content = CWRGenerator.generate_cwr(job.result)
        
        filename = f"CW_{job.file_name.rsplit('.', 1)[0]}_{job_id[:8]}.V21"
        
        return Response(
            content=cwr_content,
            # Validation: CWR is often treated as text/plain or application/octet-stream
            media_type="text/plain", 
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"CWR export failed for job {job_id}: {e}")
        raise HTTPException(status_code=500, detail=str(e))
